d2/src/main/java/com/linkedin/d2/balancer/clients/DegraderTrackerClient.java

```diff
@@ -15,173 +15,32 @@
 */
 package com.linkedin.d2.balancer.clients;
 
-import java.net.URI;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.regex.Pattern;
-
-import com.linkedin.d2.balancer.properties.PartitionData;
-import com.linkedin.d2.balancer.properties.PropertyKeys;
-import com.linkedin.r2.transport.common.bridge.client.TransportClient;
-import com.linkedin.util.clock.Clock;
-import com.linkedin.util.clock.SystemClock;
+import com.linkedin.util.degrader.CallTracker;
 import com.linkedin.util.degrader.Degrader;
 import com.linkedin.util.degrader.DegraderControl;
-import com.linkedin.util.degrader.DegraderImpl;
 
 /**
- * Used by {@link com.linkedin.d2.balancer.strategies.degrader.DegraderLoadBalancerStrategyV3}
- * with added {@link Degrader} logic.
+ * {@link TrackerClient} that contains additional methods needed for degrader strategy.
+ *
+ * @see com.linkedin.d2.balancer.strategies.degrader.DegraderLoadBalancerStrategyV3
  */
-public class DegraderTrackerClient extends TrackerClientImpl
+public interface DegraderTrackerClient extends TrackerClient
```

Comment:
This is much cleaner -- I can immediately see what DegraderTrackerClient is.

However you still need to cast between DegraderTrackerClient and TrackerClient. Would it possible to move these definitions to TrackerClient interface, while define their 'default' behavior to 'return null'? Yeah for old callTracker you cannot get degrader. In this way we can see a clean picture what TrackerClient provides.

Reply:
I'd prefer not to since to me Degrader logic is old and will no longer be used going forward. By separating DegraderTrackerClient from TrackerClient we clean up the API. Eventually our goal should be to deprecate/remove Degrader entirely.

Reply:
So that's why I was confused -- new relativeStrategy will only depend on new TrackerClient implementation without Degrader? If that's the case why we need this refactoring of DegraderTrackerClient? If we're going to keep degrader interface but have a different degraderImpl, it does make sense to me to move them into TrackerClient.

Reply:
Yes relative strategy will not use any degrader components.

We do this refactoring to clean it up. We don't want any old degrader logic in the new TrackerClient. It just adds complexity to the class for features that we don't want anymore.

Reply:
Thanks for the clarification. I thought DegraderTrackerClient will be kept. If that's the case, it's ok we do not have this new interface as it'll be gone anyway.